src/games/strategy/triplea/ui/MapData.java

```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
   }
 
   /**
+   * Find the best rectangle inside the territory to place the name in. Finds the rectangle
+   * that can fit the name, that is the closest to the vertical center, and has a large width at
+   * that location. If there isn't any rectangles that can fit the name then default back to the
+   * bounding rectangle.
+   */
+  public Rectangle getBestTerritoryNameRect(final Territory t, final FontMetrics fm) {
+    final Rectangle territoryBounds = getBoundingRect(t);
+    Rectangle maxRectangle = territoryBounds;
+    final int minX = territoryBounds.x;
+    final int minY = territoryBounds.y;
+    final int maxX = minX + territoryBounds.width;
+    final int maxY = minY + territoryBounds.height;
+    final int centerY = minY + territoryBounds.height / 2;
+    final List<Polygon> polygons = m_polys.get(t.getName());
+    final int minWidth = fm.stringWidth(t.getName());
+    final int increment = fm.getAscent();
+    int maxScore = 0;
+    for (int startX = minX; startX < maxX - increment; startX += increment) {
+      for (int startY = minY; startY < maxY - increment; startY += increment) {
+        for (int endX = maxX; endX > startX; endX -= increment) {
+          final Rectangle current = new Rectangle(startX, startY, endX - startX, increment);
+          final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));
+          final int score = verticalFromEdge * current.width;
+          if (current.width > minWidth && score > maxScore) {
+            boolean isContained = false;
+            for (final Polygon polygon : polygons) {
+              if (polygon.contains(current)) {
```

Comment:
The inner most loop is only dependent on Y (`startY`), in:
` final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));`

As Y grows, the value of `veritcalFromEdge` only shrinks  until  this is true: `startY+increment > 2 * centerY`
It does appear we would be out of the rectangle if that were ever true.
This means that the first iteration of the Y loop will find a max value, and same for the X loops.